File: dev_assistant/dev_assistant/report/mandatory_field_analytics/mandatory_field_analytics.js
// Copyright (c) 2024, Swapnil and contributors
// For license information, please see license.txt

frappe.query_reports["Mandatory Field Analytics"] = {
	"filters": [
		{
			"fieldname": "document_type",
			"label": __("Document Type"),
			"fieldtype": "Data"
		},
		{
			"fieldname": "enabled",
			"label": __("Status"),
			"fieldtype": "Select",
			"options": "\nActive\nDisabled"
		}
	],

	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (column.fieldname == "priority") {
			const priority_colors = {
				"Critical": "red",
				"High": "orange",
				"Medium": "yellow",
				"Low": "blue"
			};
			const color = priority_colors[data.priority] || "gray";
			value = `<span class="indicator-pill ${color}">${value}</span>`;
		}

		if (column.fieldname == "enabled") {
			if (data.enabled === "Active") {
				value = `<span class="indicator-pill green">${value}</span>`;
			} else {
				value = `<span class="indicator-pill gray">${value}</span>`;
			}
		}

		if (column.fieldname == "condition_count" || column.fieldname == "field_count") {
			if (parseInt(value) > 0) {
				value = `<strong>${value}</strong>`;
			}
		}

		return value;
	},

	"onload": function(report) {
		// Add custom buttons
		report.page.add_inner_button(__("Refresh"), function() {
			report.refresh();
		});

		report.page.add_inner_button(__("Export All"), function() {
			export_all_configurations();
		});

		report.page.add_inner_button(__("Create Rule"), function() {
			frappe.new_doc("Mandatory Field Controller");
		});

		report.page.add_inner_button(__("View List"), function() {
			frappe.set_route("List", "Mandatory Field Controller");
		});

		// Add help text
		add_help_section(report);
	},

	// Removed after_datatable_render to avoid conflicts

	"get_datatable_options": function(options) {
		return Object.assign(options, {
			checkboxColumn: true,
			dynamicRowHeight: true
		});
	}
};

function export_all_configurations() {
	frappe.call({
		method: "dev_assistant.dev_assistant.doctype.mandatory_field_controller.mandatory_field_controller.export_all_configurations",
		callback: function(r) {
			if (r.message) {
				const json_str = JSON.stringify(r.message, null, 2);
				const blob = new Blob([json_str], { type: "application/json" });
				const url = URL.createObjectURL(blob);
				const a = document.createElement("a");
				a.href = url;
				a.download = `mandatory_field_configurations_${frappe.datetime.nowdate()}.json`;
				a.click();

				frappe.show_alert({
					message: __("Configurations exported successfully"),
					indicator: "green"
				});
			}
		}
	});
}

function add_help_section(report) {
	// Simplified help section to avoid conflicts
	frappe.show_alert({
		message: __('Mandatory Field Analytics Report loaded successfully. Use filters to analyze your validation rules.'),
		indicator: 'blue'
	});
}

function show_empty_state(datatable_obj) {
	// Create a nice empty state
	const empty_html = `
		<div class="empty-state" style="text-align: center; padding: 60px 20px;">
			<div style="font-size: 48px; color: #d1d8dd; margin-bottom: 20px;">
				📋
			</div>
			<h3 style="color: #555; margin-bottom: 15px;">No Mandatory Field Rules Found</h3>
			<p style="color: #888; margin-bottom: 30px;">
				Get started by creating your first mandatory field validation rule.<br>
				Use rules to ensure important fields are filled before saving documents.
			</p>
			<div style="margin-bottom: 20px;">
				<button class="btn btn-primary btn-create-rule" style="margin-right: 10px;">
					<i class="fa fa-plus"></i> Create Your First Rule
				</button>
				<button class="btn btn-default btn-view-docs">
					<i class="fa fa-book"></i> View Documentation
				</button>
			</div>
			<div class="quick-start-tips" style="background: #f8f9fa; padding: 20px; border-radius: 8px; margin-top: 30px; text-align: left; max-width: 600px; margin-left: auto; margin-right: auto;">
				<h5>Quick Start Tips:</h5>
				<ul style="margin: 0;">
					<li>Choose a document type (e.g., Sales Order, Customer)</li>
					<li>Select fields that must be filled</li>
					<li>Set conditions when validation should apply</li>
					<li>Test with sample data before enabling</li>
				</ul>
			</div>
		</div>
	`;

	// Replace the datatable content
	$(datatable_obj.wrapper).html(empty_html);

	// Add click handlers
	$(datatable_obj.wrapper).find('.btn-create-rule').on('click', function() {
		frappe.new_doc("Mandatory Field Controller");
	});

	$(datatable_obj.wrapper).find('.btn-view-docs').on('click', function() {
		frappe.msgprint({
			title: __('Getting Started'),
			message: `
				<h5>Creating Mandatory Field Rules</h5>
				<ol>
					<li><strong>Select Document Type:</strong> Choose which form/document to validate</li>
					<li><strong>Add Conditions:</strong> Define when validation should apply (optional)</li>
					<li><strong>Select Fields:</strong> Choose which fields must be filled</li>
					<li><strong>Test & Enable:</strong> Test with sample data, then enable the rule</li>
				</ol>
				<p><strong>Pro Tip:</strong> Start with simple rules on important documents like Sales Orders or Customers.</p>
			`,
			indicator: 'blue'
		});
	});
}

File: dev_assistant/dev_assistant/report/mandatory_field_analytics/mandatory_field_analytics.py
# Copyright (c) 2024, Swapnil and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from datetime import datetime, timedelta


def execute(filters=None):
	"""Generate Mandatory Field Analytics Report."""
	columns = get_columns()
	data = get_data(filters)
	charts = get_charts(data)
	summary = get_summary(data)

	return columns, data, None, charts, summary


def get_columns():
	"""Define report columns."""
	return [
		{
			"label": _("Controller Name"),
			"fieldname": "name",
			"fieldtype": "Data",
			"width": 200
		},
		{
			"label": _("Document Type"),
			"fieldname": "document_type",
			"fieldtype": "Data",
			"width": 150
		},
		{
			"label": _("Priority"),
			"fieldname": "priority",
			"fieldtype": "Select",
			"options": "Low\nMedium\nHigh\nCritical",
			"width": 100
		},
		{
			"label": _("Execution Mode"),
			"fieldname": "execution_mode",
			"fieldtype": "Data",
			"width": 120
		},
		{
			"label": _("Status"),
			"fieldname": "enabled",
			"fieldtype": "Data",
			"width": 80
		},
		{
			"label": _("Conditions"),
			"fieldname": "condition_count",
			"fieldtype": "Int",
			"width": 90
		},
		{
			"label": _("Mandatory Fields"),
			"fieldname": "field_count",
			"fieldtype": "Int",
			"width": 120
		},
		{
			"label": _("Created By"),
			"fieldname": "owner",
			"fieldtype": "Data",
			"width": 150
		},
		{
			"label": _("Created On"),
			"fieldname": "creation",
			"fieldtype": "Datetime",
			"width": 150
		},
		{
			"label": _("Last Modified"),
			"fieldname": "modified",
			"fieldtype": "Datetime",
			"width": 150
		}
	]


def get_data(filters):
	"""Fetch report data based on filters."""
	conditions = build_conditions(filters)

	data = frappe.db.sql("""
		SELECT
			mfc.name,
			mfc.document_type,
			mfc.priority,
			mfc.execution_mode,
			IF(mfc.enabled, 'Active', 'Disabled') as enabled,
			(SELECT COUNT(*) FROM `tabMandatory Field Condition` WHERE parent = mfc.name) as condition_count,
			(SELECT COUNT(*) FROM `tabMandatory Field Detail` WHERE parent = mfc.name) as field_count,
			mfc.owner,
			mfc.creation,
			mfc.modified
		FROM
			`tabMandatory Field Controller` mfc
		WHERE
			{conditions}
		ORDER BY
			mfc.priority DESC, mfc.modified DESC
	""".format(conditions=conditions), as_dict=1)

	return data


def build_conditions(filters):
	"""Build SQL conditions from filters."""
	conditions = ["1=1"]

	if filters:
		if filters.get("document_type"):
			conditions.append(f"mfc.document_type = '{filters['document_type']}'")

		if filters.get("priority"):
			conditions.append(f"mfc.priority = '{filters['priority']}'")

		if filters.get("execution_mode"):
			conditions.append(f"mfc.execution_mode = '{filters['execution_mode']}'")

		if filters.get("enabled"):
			enabled_value = 1 if filters["enabled"] == "Active" else 0
			conditions.append(f"mfc.enabled = {enabled_value}")

		if filters.get("owner"):
			conditions.append(f"mfc.owner = '{filters['owner']}'")

		if filters.get("from_date"):
			conditions.append(f"DATE(mfc.creation) >= '{filters['from_date']}'")

		if filters.get("to_date"):
			conditions.append(f"DATE(mfc.creation) <= '{filters['to_date']}'")

	return " AND ".join(conditions)


def get_charts(data):
	"""Generate chart configurations."""
	charts = []

	# Priority Distribution Chart
	priority_data = {}
	for row in data:
		priority = row.get("priority", "Not Set")
		priority_data[priority] = priority_data.get(priority, 0) + 1

	charts.append({
		"title": _("Rules by Priority"),
		"data": {
			"labels": list(priority_data.keys()),
			"datasets": [{
				"values": list(priority_data.values())
			}]
		},
		"type": "donut",
		"height": 300
	})

	# Document Type Distribution
	doctype_data = {}
	for row in data:
		doctype = row.get("document_type", "Not Set")
		doctype_data[doctype] = doctype_data.get(doctype, 0) + 1

	# Get top 10 doctypes
	sorted_doctypes = sorted(doctype_data.items(), key=lambda x: x[1], reverse=True)[:10]

	if sorted_doctypes:
		charts.append({
			"title": _("Top 10 Document Types"),
			"data": {
				"labels": [d[0] for d in sorted_doctypes],
				"datasets": [{
					"name": "Rules",
					"values": [d[1] for d in sorted_doctypes]
				}]
			},
			"type": "bar",
			"height": 300
		})

	# Timeline Chart - Rules created over time
	timeline_data = {}
	for row in data:
		date_key = row.get("creation").date() if row.get("creation") else None
		if date_key:
			timeline_data[str(date_key)] = timeline_data.get(str(date_key), 0) + 1

	if timeline_data:
		sorted_timeline = sorted(timeline_data.items())
		charts.append({
			"title": _("Rules Created Over Time"),
			"data": {
				"labels": [d[0] for d in sorted_timeline],
				"datasets": [{
					"name": "Rules Created",
					"values": [d[1] for d in sorted_timeline]
				}]
			},
			"type": "line",
			"height": 300
		})

	return charts


def get_summary(data):
	"""Generate summary statistics."""
	if not data:
		# Show helpful cards when no data
		return [
			{
				"value": 0,
				"label": _("Total Rules"),
				"datatype": "Int",
				"color": "blue"
			},
			{
				"value": "Get Started",
				"label": _("Create your first Mandatory Field Rule"),
				"datatype": "Data",
				"color": "green"
			},
			{
				"value": "Documentation",
				"label": _("Check help for setup guide"),
				"datatype": "Data",
				"color": "orange"
			}
		]

	total_rules = len(data)
	active_rules = len([d for d in data if d.get("enabled") == "Active"])
	critical_rules = len([d for d in data if d.get("priority") == "Critical"])

	# Calculate average conditions and fields
	total_conditions = sum([d.get("condition_count", 0) for d in data])
	total_fields = sum([d.get("field_count", 0) for d in data])
	avg_conditions = round(total_conditions / total_rules, 1) if total_rules else 0
	avg_fields = round(total_fields / total_rules, 1) if total_rules else 0

	# Get unique document types
	unique_doctypes = len(set([d.get("document_type") for d in data if d.get("document_type")]))

	summary = [
		{
			"value": total_rules,
			"label": _("Total Rules"),
			"datatype": "Int",
			"color": "blue"
		},
		{
			"value": active_rules,
			"label": _("Active Rules"),
			"datatype": "Int",
			"color": "green"
		},
		{
			"value": critical_rules,
			"label": _("Critical Priority"),
			"datatype": "Int",
			"color": "red"
		},
		{
			"value": unique_doctypes,
			"label": _("Document Types"),
			"datatype": "Int",
			"color": "orange"
		},
		{
			"value": avg_conditions,
			"label": _("Avg Conditions/Rule"),
			"datatype": "Float",
			"color": "purple"
		},
		{
			"value": avg_fields,
			"label": _("Avg Fields/Rule"),
			"datatype": "Float",
			"color": "yellow"
		}
	]

	return summary